mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala

```diff
@@ -62,19 +77,60 @@ class RFormula(override val uid: String)
   /** @group getParam */
   def getFormula: String = $(formula)
 
-  /** @group getParam */
-  def setFeaturesCol(value: String): this.type = set(featuresCol, value)
+  override def fit(dataset: DataFrame): RFormulaModel = {
+    require(parsedFormula.isDefined, "Must call setFormula() first.")
+    val factorLevels = parsedFormula.get.terms.flatMap { term =>
+      dataset.schema(term) match {
+        case column if column.dataType == StringType =>
+          val idxTerm = term + "_idx_" + uid
+          val indexer = new StringIndexer().setInputCol(term).setOutputCol(idxTerm)
+          Some(term -> indexer.fit(dataset))
+        case _ =>
+          None
```

Comment:
It might be simpler to construct the entire preprocessing pipeline in `fit`, which includes `StringIndexer`s, `OneHotEncoder`, and `VectorAssembler`. Then call `fit` on the pipeline and pass the `PipelineModel` to `RFormulaModel`. We might add `StringVectorizer` to combine `StringIndexer` and `OneHotEncoder` in the future.

I'm a little worried about the generated feature names. But we could address this issue separately.